src/pyrokinetics/equilibrium/imas.py

```diff
@@ -0,0 +1,237 @@
+    are included.
+
+    It is not possible to determine the coordinate system used by an IMAS file from
+    its own data alone. By default, we assume that the toroidal angle :math:`\phi`
+    increases in an anti-clockwise direction when the tokamak is viewed from above. If
+    the IMAS file originates from a code that uses the opposite convention, the
+    user should set ``clockwise_phi`` to ``True``. Alternatively, if the COCOS
+    convention of the IMAS file is known, this should be supplied to the optional
+    ``cocos`` argument.
+
+    See Also
+    --------
+    Equilibrium: Class representing a global tokamak equilibrium.
+    read_equilibrium: Read an equilibrium file, return an ``Equilibrium``.
+    """
+
+    def read_from_file(
+        self,
+        filename: PathLike,
+        time: Optional[float] = None,
+        time_index: Optional[int] = None,
+        psi_n_lcfs: float = 1.0,
+        clockwise_phi: bool = False,
+        cocos: Optional[int] = None,
+    ) -> Equilibrium:
+        r"""
+        Read in IMAS file and populate Equilibrium object. Should not be invoked
+        directly; users should instead use ``read_equilibrium``.
+
+        Parameters
+        ----------
+        filename: PathLike
+            Location of the IMAS file on disk.
+        time: Optional[float]
+            The time, in seconds, at which equilibrium data should be taken. Data will
+            be drawn from the time closest to the provided value. Users should only
+            provide one of ``time`` or ``time_index``. If neither is provided, data is
+            drawn at the last time stamp.
+        time_index: Optional[int]
+            As an alternative to providing the time directly, users may provide the
+            index of the desired time stamp.
+        psi_n_lcfs: float, default 1.0
+            Adjust which flux surface we consider to be the last closed flux surface
+            (LCFS). Should take a value between 0.0 and 1.0 inclusive.
+        clockwise_phi: bool, default False
+            Determines whether the :math:`\phi` grid increases clockwise or
+            anti-clockwise when the tokamak is viewed from above.
+        cocos: Optional[int]
+            If set, asserts that the GEQDSK file follows that COCOS convention, and
+            neither ``clockwise_phi`` nor the file contents will be used to identify
+            the actual convention in use. The resulting Equilibrium is always converted
+            to COCOS 11.
+
+        Returns
+        -------
+        Equilibrium
+        """
+        # Define some units to use later
+        psi_units = units.weber / units.radian
+        F_units = units.meter * units.tesla
+
+        if time_index is not None and time is not None:
+            raise RuntimeError("Cannot set both 'time' and 'time_index'")
+
+        with h5py.File(filename, "r") as raw_file:
+            data = raw_file["equilibrium"]
```

Comment:
Could un-indent after these two lines?

Reply:
Seems like I can't do that as the file and datatype get closed...

Here is what I get in the indent and outside the indent

<HDF5 group "/equilibrium" (113 members)>
<Closed HDF5 group>

The second doesn't allow me to read any data

Reply:
Ah right, of course, sorry! Yeah, you'd need to read all the data up-front, so maybe not then